src/Microsoft.AspNetCore.Blazor.Razor.Extensions/TrimWhitespacePass.cs

```diff
@@ -0,0 +1,100 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.AspNetCore.Razor.Language.Intermediate;
+using System;
+
+namespace Microsoft.AspNetCore.Blazor.Razor
+{
+    internal class TrimWhitespacePass : IntermediateNodePassBase, IRazorDirectiveClassifierPass
+    {
+        protected override void ExecuteCore(RazorCodeDocument codeDocument, DocumentIntermediateNode documentNode)
+        {
+            if (codeDocument == null)
+            {
+                throw new ArgumentNullException(nameof(codeDocument));
+            }
+
+            if (documentNode == null)
+            {
+                throw new ArgumentNullException(nameof(documentNode));
+            }
+
+            // There's no benefit running the whitespace trimmer during design-time builds
+            if (!documentNode.Options.DesignTime)
+            {
+                var method = documentNode.FindPrimaryMethod();
+                if (method != null)
+                {
+                    RemoveContiguousWhitespace(method.Children, TraversalDirection.Forwards);
+                    RemoveContiguousWhitespace(method.Children, TraversalDirection.Backwards);
+                }
+            }
+        }
+
+        private static void RemoveContiguousWhitespace(IntermediateNodeCollection nodes, TraversalDirection direction)
+        {
+            var position = direction == TraversalDirection.Forwards ? 0 : nodes.Count - 1;
+            while (position >= 0 && position < nodes.Count)
+            {
+                var node = nodes[position];
+                bool shouldRemoveNode;
+                bool shouldContinueIteration;
+
+                switch (node)
+                {
+                    case IntermediateToken intermediateToken:
+                        shouldRemoveNode = string.IsNullOrWhiteSpace(intermediateToken.Content);
+                        shouldContinueIteration = shouldRemoveNode;
+                        break;
+
+                    case HtmlContentIntermediateNode htmlContentIntermediateNode:
+                        RemoveContiguousWhitespace(htmlContentIntermediateNode.Children, direction);
+                        shouldRemoveNode = htmlContentIntermediateNode.Children.Count == 0;
+                        shouldContinueIteration = shouldRemoveNode;
+                        break;
+
+                    case CSharpExpressionIntermediateNode _:
+                    case TagHelperIntermediateNode _:
```

Comment:
What about a case with child content? Is that covered?

```
<MyComponent>


  <button onclick="@foo">

      @bar

  </button>


</MyComponent>
```

In each of these cases, this will be lowered to a form like:
```
Method
  TagHelper
    ...
    TagHelperBody
      ...
      HtmlContent
      ...
```

Reply:
In that scenario the algorithm terminates as soon as it sees the `TagHelperIntermediateNode` - the contents of that node don't matter. We're only trying to trim the top-level leading and trailing whitespace in the entire document.

Reply:
OK cool 👍